src/abstractdatatable/join.jl

```diff
@@ -3,217 +3,170 @@
-    for i in 1:(max_groups + 1)
-        lc = left_count[i]
-        rc = right_count[i]
-        if rc == 0
-            for j in 1:lc
-                leftonly_indexer[lposition + j] = left_pos + j
-            end
-            lposition += lc
-        elseif lc == 0
-            for j in 1:rc
-                rightonly_indexer[rposition + j] = right_pos + j
-            end
-            rposition += rc
-        else
-            for j in 1:lc
-                offset = tposition + (j-1) * rc
-                for k in 1:rc
-                    left_indexer[offset + k] = left_pos + j
-                    right_indexer[offset + k] = right_pos + k
-                end
-            end
-            tposition += lc * rc
-        end
-        left_pos += lc
-        right_pos += rc
-    end
+DataTableJoiner{DT1<:AbstractDataTable, DT2<:AbstractDataTable}(dtl::DT1, dtr::DT2, on::Union{Symbol,Vector{Symbol}}) =
+    DataTableJoiner{DT1,DT2}(dtl, dtr, on)
 
-    ## (left_sorter, left_indexer, leftonly_indexer,
-    ##  right_sorter, right_indexer, rightonly_indexer)
-    (left_sorter[left_indexer], left_sorter[leftonly_indexer],
-     right_sorter[right_indexer], right_sorter[rightonly_indexer])
+# helper map between the row indices in original and joined table
+immutable RowIndexMap
+    "row indices in the original table"
+    orig::Vector{Int}
+    "row indices in the resulting joined table"
+    join::Vector{Int}
 end
 
-function sharepools{S,N}(v1::Union{CategoricalArray{S,N}, NullableCategoricalArray{S,N}},
-                         v2::Union{CategoricalArray{S,N}, NullableCategoricalArray{S,N}},
-                         index::Vector{S},
-                         R)
-    tidx1 = convert(Vector{R}, indexin(CategoricalArrays.index(v1.pool), index))
-    tidx2 = convert(Vector{R}, indexin(CategoricalArrays.index(v2.pool), index))
-    refs1 = zeros(R, length(v1))
-    refs2 = zeros(R, length(v2))
-    for i in 1:length(refs1)
-        if v1.refs[i] != 0
-            refs1[i] = tidx1[v1.refs[i]]
-        end
+Base.length(x::RowIndexMap) = length(x.orig)
+
+# composes the joined data table using the maps between the left and right
+# table rows and the indices of rows in the result
+function compose_joined_table(joiner::DataTableJoiner,
+                              left_ixs::RowIndexMap, leftonly_ixs::RowIndexMap,
+                              right_ixs::RowIndexMap, rightonly_ixs::RowIndexMap)
+    @assert length(left_ixs) == length(right_ixs)
+    # compose left half of the result taking all left columns
+    all_orig_left_ixs = vcat(left_ixs.orig, leftonly_ixs.orig)
+    if length(leftonly_ixs) > 0
+        # permute the indices to restore left table rows order
+        all_orig_left_ixs[vcat(left_ixs.join, leftonly_ixs.join)] = all_orig_left_ixs
+    end
+    ril = length(right_ixs)
+    loil = length(leftonly_ixs)
+    roil = length(rightonly_ixs)
+    left_dt = DataTable(Any[resize!(col[all_orig_left_ixs], length(all_orig_left_ixs)+roil)
```

Comment:
As I wrote sometime earlier, memory allocation can happen twice here: once for `col[all_orig_left_ixs]`, once for `resize!()`. It might be worth declaring some internal method like `permpad(col, ixs, npad=0)` that would allocate an array of `(length(ixs)+npad)` size and then copy `col[ixs]` into it in a for loop.